File: benchmarks/levenshtein/levenshtein.rb
def levenshtein(s, t)
  return t.length if s.empty?
  return s.length if t.empty?

  n = t.length
  prev = (0..n).to_a
  curr = Array.new(n + 1, 0)

  s.each_char.with_index do |sc, i|
    curr[0] = i + 1
    t.each_char.with_index do |tc, j|
      cost = sc == tc ? 0 : 1
      ins = curr[j] + 1
      del = prev[j + 1] + 1
      sub = prev[j] + cost
      m = ins < del ? ins : del
      curr[j + 1] = sub < m ? sub : m
    end
    prev, curr = curr, prev
  end

  prev[n]
end

s = "the quick brown fox jumps over the lazy dog" * 20
t = "the slow brown fox leaps over the lazy cat" * 20
puts levenshtein(s, t)

File: benchmarks/levenshtein/levenshtein.rs
fn levenshtein(s: &[u8], t: &[u8]) -> usize {
    let (m, n) = (s.len(), t.len());
    if m == 0 {
        return n;
    }
    if n == 0 {
        return m;
    }

    let mut prev: Vec<usize> = (0..=n).collect();
    let mut curr: Vec<usize> = vec![0; n + 1];

    for i in 0..m {
        curr[0] = i + 1;
        for j in 0..n {
            let cost = if s[i] == t[j] { 0 } else { 1 };
            let ins = curr[j] + 1;
            let del = prev[j + 1] + 1;
            let sub = prev[j] + cost;
            curr[j + 1] = ins.min(del).min(sub);
        }
        std::mem::swap(&mut prev, &mut curr);
    }

    prev[n]
}

fn main() {
    let s = "the quick brown fox jumps over the lazy dog".repeat(20);
    let t = "the slow brown fox leaps over the lazy cat".repeat(20);
    println!("{}", levenshtein(s.as_bytes(), t.as_bytes()));
}
